### ISSUE TITLE
Bug in unique_roots in scipy.signal.signaltools.py for roots with same magnitude

### ISSUE THREAD

**Issue by chipmuenk** [scipy/scipy#4915]
In my opinion, unique_roots() from scipy.signal.signaltools.py fails for multiple roots with equal magnitude. Simple example:

Generate four double roots on the unit circle with  
```
np.roots(np.convolve(np.ones(5), np.ones(5)))
```
Result: 
```
(array([-0.809+0.5878j, -0.809-0.5878j,  0.309+0.9511j,  0.309-0.9511j,
         0.309+0.9511j,  0.309-0.9511j, -0.809+0.5878j, -0.809-0.5878j]),
 array([1, 1, 1, 1, 1, 1, 1, 1]))
```

![roots_bug](https://cloud.githubusercontent.com/assets/4620020/7828802/d481e99c-0438-11e5-97cc-100be67f7aab.png)

Expected: (
```
array([-0.809-0.5878j, -0.809+0.5878j, 0.309-0.9511j,  0.309+0.9511j]),
array([2, 2, 2, 2]))
```

![roots_correct](https://cloud.githubusercontent.com/assets/4620020/7828813/edec4ca6-0438-11e5-8023-a77adbd5cb23.png)

The root cause IMHO is the concept of trying to sort the roots by magnitude before grouping them which fails for multiple roots with identical radius.

As a replacement for my purposes, I've re-written unique_roots() in

https://github.com/chipmuenk/pyFDA/blob/master/pyFDA/pyfda_lib.py

which calculates the distance between _all_ roots and groups those within a given distance eps. As a drawback, the algorithm is O^2 and hence slower than the original algorithm (~ factor 20 for 500 double roots). For my cases it works well enough although I'm not quite clear about what _should_ happen when you have a cluster of roots with distances between individual poles < eps but with an overall radius > eps.

I've also written some unittest testcases which I could provide (although some code clean-up would be required first ...)

**Update**
Endolith, thanx for the comment: You might be right - one could sort the roots according to their real part and only calculate the (euclidian) distances between those roots where the difference between the real parts is less than the specified tolerance. I'm not sure though, which number of roots is the break-even point. I'll try and think about that.

In the meantime I've improved speed and memory consumption of the algorithm and added unittests and speed test under https://github.com/chipmuenk/pyFDA/tree/master/pyFDA/test

Worst case is btw for lots of _single_ complex roots (new algorithm is slower by a factor of 25 for 500 single complex roots). Computation time is growing a little stronger than ~N but not ~N^2.

My version of unique_roots also handles empty, scalar and nan inputs without throwing an error. Do you think this makes sense?

**Comment by endolith** [comment]
> four double roots on the unit circle

Yes this also looks like a bug to me.

> the algorithm is O^2 and hence slower than the original algorithm 

I didn't look through your algorithm yet, but if you pre-sort the roots with a fast algorithm, does it speed up processing in most cases?  I noticed that happened when I was working on _cplxreal https://github.com/scipy/scipy/blob/master/scipy/signal/filter_design.py#L409

> I'm not quite clear about what should happen when you have a cluster of roots with distances between individual poles < eps but with an overall radius > eps.

In my opinion, this should raise an exception, because there's no way of knowing what the user intended the grouping to be an they need to refine something to get the desired results.

**Comment by chipmuenk** [comment]
I updated the code and added some comments - see above.

**Comment by rgommers** [comment]
Behavior still the same:
```
In [10]: signal.unique_roots(np.roots(np.convolve(np.ones(5), np.ones(5))))                                                            
Out[10]: 
(array([-0.80901698+0.58778524j, -0.80901698-0.58778524j,
         0.309017  +0.95105651j,  0.309017  -0.95105651j,
         0.30901699+0.95105652j,  0.30901699-0.95105652j,
        -0.80901701+0.58778526j, -0.80901701-0.58778526j]),
 array([1, 1, 1, 1, 1, 1, 1, 1]))
```

**Comment by nmayorov** [comment]
Working on a proper versions of `unique_roots`.

**Comment by larsoner** [comment]
This should have been fixed by #10961
